designs/0006-labels.md

```diff
@@ -0,0 +1,289 @@
+
+### Decision 5: How will we get labels that Karpenter doesn't currently write set on nodes?
+
+#### Option A: Wait for Machine API migration
+
+Migrating to the machine API will automatically get all missing labels set on VMs allocated via the machine API.
+
+#### Option B: Add some/all labels to our current label writing mechanism
+
+There are some labels that Karpenter sets but that aren't schedulable, defined at [AddAgentBakerGeneratedLabels](https://github.com/Azure/karpenter-provider-azure/blob/main/pkg/providers/imagefamily/labels/labels.go#L23).
+This has already been accounted for in the tables above. We could add more labels there.
+
+#### Conclusion: Wait for machine API for complete parity
+
+Machine API is not that far away, and we already write most critical labels. Some additional labels such as `kubernetes.azure.com/os-sku` and `kubernetes.azure.com/fips_enabled` will be enabled for scheduling
+(and thus be written to the nodes) based on decision 4, but other labels and full label parity will come with the machine API migration.
+
+## FAQ
+
+### How do the labels get onto the Node?
+
+* Karpenter core writes all NodeClaim labels to the node, see: [syncNode](https://github.com/kubernetes-sigs/karpenter/blob/38e728c99b530f660f685d53d01f2e9ec9696668/pkg/controllers/nodeclaim/lifecycle/registration.go#L128).
+* Karpenter core translates most requirements from the NodeClaim spec into labels, see [PopulateNodeClaimDetails](https://github.com/kubernetes-sigs/karpenter/blob/38e728c99b530f660f685d53d01f2e9ec9696668/pkg/controllers/nodeclaim/lifecycle/launch.go#L129)
+    * Note that `.Labels()` does exclude multi-valued requirements and labels defined in `WellKnownLabels` (registered by the cloud provider). See [Labels()](https://github.com/kubernetes-sigs/karpenter/blob/38e728c99b530f660f685d53d01f2e9ec9696668/pkg/scheduling/requirements.go#L270).
+* Labels that get skipped by core (either due to `WellKnownLabels` or having multiple values) are added to the NodeClaim by AKS Karpenter provider in [vmInstanceToNodeClaim](https://github.com/Azure/karpenter-provider-azure/blob/b9c8c82edb289ac5b281c85b0851b5a0c45bc4bb/pkg/cloudprovider/cloudprovider.go#L476).
+  * Note that `WellKnownLabels` seems to serve two purposes:
+      1. It prevents Karpenter core from automatically including those labels on the node object (though note that AKS Karpenter provider ends up including most of them on the node object anyway).
+      2. It allows pods to ask for that label when the NodeClaim/requirements don't actually have it, and still have scheduling take place (I don't fully understand where/why this is needed but I can see the code for it [here](https://github.com/kubernetes-sigs/karpenter/blob/main/pkg/scheduling/requirements.go#L175))
+* Requirements come from InstanceType via [computeRequirements](https://github.com/Azure/karpenter-provider-azure/blob/1f327c6d7ac62b3a3a1ad83c0f14f95001ab4ae8/pkg/providers/instancetype/instancetype.go#L135)
```

Comment:
Narrative-wise, one would expect this to be about NodeClaim spec requirements - which [come](https://github.com/kubernetes-sigs/karpenter/blob/565c33c673d28f5f5a442182dcc14db6cb6b44ea/pkg/controllers/provisioning/scheduling/nodeclaimtemplate.go#L74-L85) from multiple sources (including NodePool requirements and labels) and don't include requirements derived from instance type? (Those get translated into labels by cloud provider ...)

Reply:
You're right, the way I wrote this is confusing because I'm mixing requirements on InstanceType and "actual requirements".

I've cleaned this up some, I think. The focus here is on labels so we're interested in the requirements on instance type and less so on the actual requirements on the NodePool.